### DISCUSSION TITLE
How to integreate your CSS in my application?

### DISCUSSION THREAD

**Discussion by buhtz** [pygments/pygments#2380] [Q&A] [answered]
Dear Team,

I'm sorry for asking this way but I couldn't find a mailing list or forum. This isn't a bug or feature request.

The HTML output of pygments need an external css file to look good. I can see in the docu that pygments do "offer" a lot of this styles. For example there is this file: https://pygments.org/_static/pygments.css on your own docu site.

I'm the developer and maintainer of [`hyperorg`](https://codeberg.org/buhtz/hyperorg) which do generate HTML files (out of org(roam) files). I want to to integrate pygments to offer good looking code blocks in that HTML files.

My `hyperorg` to generate a css file.

I need a second css file for your pygments stuff. How should I realize this and what is allowed?
My users may choose one of the available pygments styles. **Where should I take the related CSS file from?**

Can I do something like that?

```
import pathlib
import pygments
full_css_file_content = pygments.get_css(stylename='tango')
css_file_path = pathlib.Path(output_folder_choosen_by_my_user) / 'pygments.css'

with css_file_path.open('w') as handle:
    handle.write(full_css_file_content)
```

Or can I copy CSS files from your Microsoft GitHub repo and copy them to my own repository and pack/ship/release it then?

One main goal of my application is to generate HTML files that can be use on the local system without a webserver and without an internet connection. Because of that it isn't an option to load a pyments.css file via a remote URL like this.

`<link rel="stylesheet" href="https://github.com/pygments/pygments_tango.css">`

**Comment by jeanas** [comment, marked as answer]
```pycon
>>> from pygments.formatters import HtmlFormatter
>>> from pygments.styles import TangoStyle
>>> print(HtmlFormatter(style=TangoStyle).get_style_defs())
pre { line-height: 125%; }
td.linenos .normal { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
span.linenos { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
td.linenos .special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
.hll { background-color: #ffffcc }
.c { color: #8f5902; font-style: italic } /* Comment */
.err { color: #a40000; border: 1px solid #ef2929 } /* Error */
[...]
```

or use `pygmentize -S`. Cf. https://pygments.org/docs/api/#pygments.formatter.Formatter.get_style_defs

**Comment by buhtz** [comment]
Please avoid closing question until they are solved.
Please improve the description for your label "T-question" because it is unclear what the T stands for.

OK, technically it is possible.

Is it allowed from a license point of view?

**Comment by jeanas** [comment]
> Please avoid closing question until they are solved.

Uh, it is normally users who need to adapt to “local lore” rather than maintainers.

Sorry if you took it as rude, it really wasn't meant so. I just want to avoid forgetting issues open.

Mailing lists don't have a way to “close” topics and that's not a problem. Think of this “closed” issue as being such a discussion thread.

> Please improve the description for your label "T-question" because it is unclear what the T stands for.

"Type" I guess?

> Is it allowed from a license point of view?

Yes. The output is generated by Pygments, not written by the Pygments authors. It's the result of the program, not part of the program itself, and Pygments license (BSD) allows you to do whatever you want to with the output

**Comment by buhtz** [comment]
> > Please improve the description for your label "T-question" because it is unclear what the T stands for.
> 
> "Type" I guess?

:laughing: Sorry, I still don't get it. What is a "type question"? I'm just asking to improve my own maintainer skills. :smile: 

> Yes. The output is generated by Pygments, not written by the Pygments authors. It's the result of the program, not part of the program itself, and Pygments license (BSD) allows you to do whatever you want to with the output

You say there is no css file (in the repo or somewhere in the code as a big string) for each style? IT is generated from scratch every time? Fascinating!

OK, then I have an idea how to integrate that into my `hyperorg` application.

**Comment by jeanas** [comment]
> 😆 Sorry, I still don't get it. What is a "type question"? I'm just asking to improve my own maintainer skills. 😄

The type of the issue is “question”, as opposed to T-feature for example.

> IT is generated from scratch every time? 

Yes, if you look into the source code in pygments/styles/, each style class contains a dict of styles, which are in CSS-like syntax but not CSS syntax, and it is the formatter that generates a stylesheet. The LaTeX formatter, for example, does it differently than outputting CSS code

**Comment by birkenfeld** [comment]
> Mailing lists don't have a way to “close” topics and that's not a problem. Think of this “closed” issue as being such a discussion thread.

Meta: Should we maybe enable the "Discussions" tab for such questions? It is true that we don't have a mailing list/other forum dedicated to Pygments, so that would be nice and low-maintainer-effort solution.

**Comment by birkenfeld** [comment]
> Sorry, I still don't get it. What is a "type question"? I'm just asking to improve my own maintainer skills.

To go a bit more into detail here, the `X-foo` label names are a convention that works around the fact that GitHub has no further fields for issues like "type", "severity" or "topic". To make it clear which category a label belongs to (other than color, which is not very accessible), it is added as a short prefix.

I just made this a bit more consistent with our labels, adding more prefixes and updating the label descriptions to reflect.

**Comment by jeanas** [comment]
> > Mailing lists don't have a way to “close” topics and that's not a problem. Think of this “closed” issue as being such a discussion thread.
> 
> Meta: Should we maybe enable the "Discussions" tab for such questions? It is true that we don't have a mailing list/other forum dedicated to Pygments, so that would be nice and low-maintainer-effort solution.

Yes, sounds good!
